src/stan/io/reader.hpp

```diff
@@ -1380,42 +1705,161 @@ class reader {
+   * @tparam VecR Vector with an integral type.
+   * @tparam VecC Vector with an integral type.
+   * @param lb The lower bound to check all scalar values against.
+   * @param ub The upper bound to check all scalar values against.
+   * @param vec_r rows corresponding the nonzero entries.
+   * @param vec_c columns corresponding the nonzero entries.
+   * @param n Number of rows.
+   * @param m Number of columns.
+   * @return Eigen::SparseMatrix made up of the next scalars.
+   */
+  template <typename TL, typename TU, typename VecR, typename VecC,
+            typename = require_all_std_vector_vt<std::is_integral, VecR, VecC>>
+  inline sparse_matrix_t sparse_matrix_lub_constrain(const TL lb, const TU ub,
+                                                     const VecR& vec_r,
+                                                     const VecC& vec_c,
+                                                     size_t n, size_t m) {
+    stan::math::check_matching_sizes("sparse_matrix_lub_constrain", "vec_r",
+                                     vec_r, "vec_c", vec_c);
+    Eigen::SparseMatrix<T> ret_mat(n, m);
+    if (m == 0 || n == 0) {
+      return ret_mat;
+    }
+    using triplet_type = Eigen::Triplet<T>;
+    std::vector<triplet_type> triplet_list(vec_r.size());
+    for (auto i = 0; i < vec_r.size(); i++) {
+      triplet_list.emplace_back(
+          triplet_type(vec_r[i], vec_c[i], scalar_lub_constrain(lb, ub)));
+    }
+    ret_mat.setFromTriplets(triplet_list.begin(), triplet_list.end());
+    return ret_mat;
+  }
+
+  /**
+   * Return a sparse matrix of the specified dimensionality and number of
+   * nonzero elements made up of the next scalars scaled to the ub, lb, and
+   * and log probability arranged in column-major order.
+   *
+   * @tparam TL the type of the lower bound.
+   * @tparam TU the type of the upper bound.
+   * @tparam VecR Vector with an integral type.
+   * @tparam VecC Vector with an integral type.
+   * @param lb The lower bound to check all scalar values against.
+   * @param ub The upper bound to check all scalar values against.
+   * @param vec_r rows corresponding the nonzero entries.
+   * @param vec_c columns corresponding the nonzero entries.
+   * @param n Number of rows.
+   * @param m Number of columns.
+   * @param lp The log probability.
+   * @return Eigen::SparseMatrix made up of the next scalars.
+   */
+  template <typename TL, typename TU, typename VecR, typename VecC,
+            typename = require_all_std_vector_vt<std::is_integral, VecR, VecC>>
+  inline sparse_matrix_t sparse_matrix_lub_constrain(const TL lb, const TU ub,
+                                                     const VecR& vec_r,
+                                                     const VecC& vec_c,
+                                                     size_t n, size_t m,
+                                                     T& lp) {
+    stan::math::check_matching_sizes("sparse_matrix_lub_constrain", "vec_r",
```

Comment:
I really want to get rid of all this cut-and-paste code with some kind of simplifying abstraction.